src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,335 @@
+                // We can skip this update check.
+                return;
+            }
+
+            try
+            {
+                // Use 'sentinelFilePath' as the file lock.
+                // The update-check tasks started by every 'pwsh' process of the same version will compete on holding this file.
+                using (new FileStream(sentinelFilePath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None, bufferSize: 1, FileOptions.DeleteOnClose))
+                {
+                    if (File.Exists(todayDoneFilePath))
+                    {
+                        // After acquiring the file lock, it turns out a successful check has already been done for today.
+                        // Then let's skip this update check.
+                        return;
+                    }
+
+                    // Now it's guaranteed that this is the only process that reaches here.
+                    // Clean up the old '.done' file, there should be only one of it.
+                    foreach (string oldFile in Directory.EnumerateFiles(s_cacheDirectory, DoneFileNamePattern, s_enumOptions))
+                    {
+                        File.Delete(oldFile);
+                    }
+
+                    if (!parseSuccess)
+                    {
+                        // The update file is corrupted, either because more than one update files were found unexpectedly,
+                        // or because the update file name is not in the valid format.
+                        // This is **very unlikely** to happen unless the file is accidentally altered manually.
+                        // We try to recover here by cleaning up all update files.
+                        foreach (string file in Directory.EnumerateFiles(s_cacheDirectory, UpdateFileNamePattern, s_enumOptions))
+                        {
+                            File.Delete(file);
+                        }
+                    }
+
+                    // Do the real update check:
+                    //  - Send HTTP request to query for the new release/pre-release;
+                    //  - If there is a valid new release that should be reported to the user,
+                    //    create the file `update_<tag>_<publish-date>` when no `update` file exists,
+                    //    or rename the existing file to `update_<new-version>_<new-publish-date>`.
+                    SemanticVersion baselineVersion = lastUpdateVersion ?? PSVersionInfo.PSCurrentVersion;
+                    Release release = await QueryNewReleaseAsync(baselineVersion);
+
+                    if (release != null)
+                    {
+                        string newUpdateFileName = string.Format(
+                            CultureInfo.InvariantCulture,
+                            UpdateFileNameTemplate,
+                            release.TagName,
+                            release.PublishAt.Substring(0, 10));
```

Comment:
It would be good to have self-documented const name for 10.

Reply:
OK, will declare a const variable for this value.